GoogleSignIn/Sources/GIDProfileDataFetcher/Implementations/GIDProfileDataFetcher.m

```diff
@@ -0,0 +1,87 @@
+
+#ifdef SWIFT_PACKAGE
+@import AppAuth;
+@import GTMAppAuth;
+#else
+#import <AppAuth/AppAuth.h>
+#import <GTMAppAuth/GTMAppAuth.h>
+#endif
+
+NS_ASSUME_NONNULL_BEGIN
+
+// The URL template for the URL to get user info.
+static NSString *const kUserInfoURLTemplate = @"https://%@/oauth2/v3/userinfo";
+
+@implementation GIDProfileDataFetcher {
+  id<GIDHTTPFetcher> _httpFetcher;
+}
+
+- (instancetype)init {
+  GIDHTTPFetcher *httpFetcher = [[GIDHTTPFetcher alloc] init];
+  return [self initWithDataFetcher:httpFetcher];
+}
+
+- (instancetype)initWithDataFetcher: (id<GIDHTTPFetcher>)httpFetcher {
+  self = [super init];
+  if (self) {
+    _httpFetcher = httpFetcher;
+  }
+  return self;
+}
+
+- (void)fetchProfileDataWithAuthState:(OIDAuthState *)authState
+                           completion:(void (^)(GIDProfileData *_Nullable profileData,
+                                                NSError *_Nullable error))completion {
+  OIDIDToken *idToken =
+      [[OIDIDToken alloc] initWithIDTokenString:authState.lastTokenResponse.idToken];
+  // If the profile data are present in the ID token, use them.
+  if (idToken) {
+    GIDProfileData *profileData = [[GIDProfileData alloc]initWithIDToken:idToken];
+    completion(profileData, nil);
+    return;
+  }
+  // If we can't retrieve profile data from the ID token, make a userInfo request to fetch them.
+  NSString *infoString = [NSString stringWithFormat:kUserInfoURLTemplate,
+                             [GIDSignInPreferences googleUserInfoServer]];
+  NSURL *infoURL = [NSURL URLWithString:infoString];
+  NSMutableURLRequest *infoRequest = [NSMutableURLRequest requestWithURL:infoURL];
+  GTMAppAuthFetcherAuthorization *authorization =
+      [[GTMAppAuthFetcherAuthorization alloc] initWithAuthState:authState];
+  
+  [_httpFetcher fetchURLRequest:infoRequest
+                 withAuthorizer:authorization
+                    withComment:@"GIDSignIn: fetch basic profile info"
+                     completion:^(NSData *data, NSError *error) {
+    if (error) {
+      completion(nil, error);
+    } else {
+      NSError *jsonDeserializationError;
+      NSDictionary<NSString *, NSString *> *profileDict =
+        [NSJSONSerialization JSONObjectWithData:data
+                                        options:NSJSONReadingMutableContainers
+                                          error:&jsonDeserializationError];
+      if (profileDict) {
+        GIDProfileData *profileData = [[GIDProfileData alloc]
+            initWithEmail:idToken.claims[kBasicProfileEmailKey]
+                     name:profileDict[kBasicProfileNameKey]
+                givenName:profileDict[kBasicProfileGivenNameKey]
+               familyName:profileDict[kBasicProfileFamilyNameKey]
+                 imageURL:[NSURL URLWithString:profileDict[kBasicProfilePictureKey]]];
+        completion(profileData, nil);
+      }
+      else {
+        completion(nil, jsonDeserializationError);
+      }
```

Comment:
Similarly, handle the error case and return first, then handle the success case.

```
if (jsonDeserializationError) {
  completion(nil, jsonDeserializationError);
  return;
}
GIDProfileData *profileData = ...
```

Reply:
Done.